pkg/addr/addr.go

```diff
@@ -21,13 +21,12 @@ const DefaultWeight = 1
 // string, the IP address should be enclosed in square brackets followed by
 // the port.
 func PublicAddressToString(addr *l5dNetPb.TcpAddress) string {
-	var s string
-	if addr.GetIp().GetIpv6() != nil {
-		s = "[%s]:%d"
-	} else {
-		s = "%s:%d"
+	var strIP, strPort string
+	if addr != nil {
+		strIP = PublicIPToString(addr.GetIp())
+		strPort = fmt.Sprint(addr.GetPort())
```

Comment:
This condition unfortunately breaks the existing unit test. The unit test (`TestPublicAddressesToString()`) expects to see `"<nil>:0"` when the IP address is nil (or empty string). Before, we used to call `PublicIPToString` with the IP address regardless if it was nil, when we tried to convert it to a string using `net.IP(b).String()`, the byte array would be empty (since no IP is present for either condition to happen).

From the Go docs:

```
String returns the string form of the IP address ip. It returns one of 4 forms:
- "<nil>", if ip has length 0
...
```

[[ref](https://pkg.go.dev/net#IP.String)]

Now, however, we simply join the two using `JoinHostPort` and we get ":" as a result.

Here's what I'd suggest:

```suggestion
    strIP := PublicIPToString(addr.GetIp())
	strPort := strconv.Itoa(int(addr.GetPort()))
	return net.JoinHostPort(strIP, strPort)
```

1. `PublicIPToString(addr.GetIp())` will return "<nil>" or "<ip>"
2. `strconv.Itoa(int(addr.GetPort()))` we convert unsigned int to int, then we convert to string. AFAIK this operation should be **infallible**, but you (or second reviewer) might wanna double check.
3. convert to string using `JoinHostPort`. If we have a port and an IP, we'll get the value (regardless of IP class). If IP is nil and port 0, we get "<nil>:0" just as the test expects.

The function also becomes much simpler as a side effect :) 

Wdyt?